accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,234 @@
+type DeployFn func(input, deployer []byte) (common.Address, *types.Transaction, error)
+
+// depTreeDeployer is responsible for taking a dependency, deploying-and-linking its components in the proper
+// order.  A depTreeDeployer cannot be used after calling LinkAndDeploy other than to retrieve the deployment result.
+type depTreeDeployer struct {
+	deployedAddrs map[string]common.Address
+	deployerTxs   map[string]*types.Transaction
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(node *depTreeNode) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if node.overrideAddr != nil {
+		return nil
+	}
+
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, childNode := range node.children {
+		if err := d.linkAndDeploy(childNode); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := node.unlinkedCode
+	for _, child := range node.children {
+		linkAddr := d.deployedAddrs[child.pattern]
+		if child.overrideAddr != nil {
+			linkAddr = *child.overrideAddr
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+child.pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
+
+	// Finally, deploy the contract.
+	addr, tx, err := d.deploy(d.input[node.pattern], common.Hex2Bytes(deployerCode))
+	if err != nil {
+		return err
+	}
+
+	d.deployedAddrs[node.pattern] = addr
+	d.deployerTxs[node.pattern] = tx
+	return nil
+}
+
+// result returns a result for this deployment, or an error if it failed.
+func (d *depTreeDeployer) result() *DeploymentResult {
+	return &DeploymentResult{
+		Txs:   d.deployerTxs,
+		Addrs: d.deployedAddrs,
+	}
+}
+
+func newDepTreeDeployer(deploy DeployFn) *depTreeDeployer {
+	return &depTreeDeployer{
+		deploy:        deploy,
+		deployedAddrs: make(map[string]common.Address),
+		deployerTxs:   make(map[string]*types.Transaction)}
+}
+
+// LinkAndDeploy deploys a specified set of contracts and their dependent
+// libraries.  If an error occurs, only contracts which were successfully
+// deployed are returned in the result.
+func LinkAndDeploy(deployParams *DeploymentParams, deploy DeployFn) (res *DeploymentResult, err error) {
+	unlinkedContracts := make(map[string]string)
+	accumRes := &DeploymentResult{
+		Txs:   make(map[string]*types.Transaction),
+		Addrs: make(map[string]common.Address),
+	}
+	for _, meta := range deployParams.contracts {
+		unlinkedContracts[meta.Pattern] = meta.Bin[2:]
```

Comment:
Are we fully in control of the inputs here? Do we know with 100% certainty that the `meta.Bin` starts with `0x`-prefix?

Reply:
I assume that users are sourcing everything from the output of solidity abi compilation.  That's why I haven't yet added more defensive checks throughout this PR.

Reply:
I don't think the contract deployer can be empty in the ABI.  I can see no reason why that would ever be a valid result of the solidity compilation process.